File: elixir/test/bullmq/keys_test.exs
defmodule BullMQ.KeysTest do
  use ExUnit.Case, async: true

  alias BullMQ.Keys

  describe "context/2" do
    test "creates a context with prefix and name" do
      ctx = Keys.context("bull", "test-queue")
      assert ctx.prefix == "bull"
      assert ctx.name == "test-queue"
    end

    test "creates a context with default prefix" do
      ctx = Keys.context("test-queue")
      assert ctx.prefix == "bull"
      assert ctx.name == "test-queue"
    end
  end

  describe "key generation" do
    setup do
      {:ok, ctx: Keys.context("bull", "emails")}
    end

    test "generates wait key", %{ctx: ctx} do
      assert Keys.wait(ctx) == "bull:emails:wait"
    end

    test "generates active key", %{ctx: ctx} do
      assert Keys.active(ctx) == "bull:emails:active"
    end

    test "generates delayed key", %{ctx: ctx} do
      assert Keys.delayed(ctx) == "bull:emails:delayed"
    end

    test "generates prioritized key", %{ctx: ctx} do
      assert Keys.prioritized(ctx) == "bull:emails:prioritized"
    end

    test "generates completed key", %{ctx: ctx} do
      assert Keys.completed(ctx) == "bull:emails:completed"
    end

    test "generates failed key", %{ctx: ctx} do
      assert Keys.failed(ctx) == "bull:emails:failed"
    end

    test "generates stalled key", %{ctx: ctx} do
      assert Keys.stalled(ctx) == "bull:emails:stalled"
    end

    test "generates limiter key", %{ctx: ctx} do
      assert Keys.limiter(ctx) == "bull:emails:limiter"
    end

    test "generates events key", %{ctx: ctx} do
      assert Keys.events(ctx) == "bull:emails:events"
    end

    test "generates marker key", %{ctx: ctx} do
      assert Keys.marker(ctx) == "bull:emails:marker"
    end

    test "generates meta key", %{ctx: ctx} do
      assert Keys.meta(ctx) == "bull:emails:meta"
    end

    test "generates id key", %{ctx: ctx} do
      assert Keys.id(ctx) == "bull:emails:id"
    end

    test "generates waiting-children key", %{ctx: ctx} do
      assert Keys.waiting_children(ctx) == "bull:emails:waiting-children"
    end

    test "generates repeat key", %{ctx: ctx} do
      assert Keys.repeat(ctx) == "bull:emails:repeat"
    end

    test "generates schedulers key", %{ctx: ctx} do
      assert Keys.schedulers(ctx) == "bull:emails:sc"
    end
  end

  describe "job-specific keys" do
    setup do
      {:ok, ctx: Keys.context("bull", "emails")}
    end

    test "generates job key", %{ctx: ctx} do
      assert Keys.job(ctx, "123") == "bull:emails:123"
    end

    test "generates lock key", %{ctx: ctx} do
      assert Keys.lock(ctx, "123") == "bull:emails:123:lock"
    end

    test "generates dependencies key", %{ctx: ctx} do
      assert Keys.dependencies(ctx, "123") == "bull:emails:123:dependencies"
    end

    test "generates processed key", %{ctx: ctx} do
      assert Keys.processed(ctx, "123") == "bull:emails:123:processed"
    end

    test "generates logs key", %{ctx: ctx} do
      assert Keys.logs(ctx, "123") == "bull:emails:123:logs"
    end
  end

  describe "key/1" do
    test "generates base key without suffix" do
      ctx = Keys.context("bull", "emails")
      assert Keys.key(ctx) == "bull:emails"
    end
  end

  describe "custom prefix" do
    test "uses custom prefix in all keys" do
      ctx = Keys.context("myapp", "notifications")

      assert Keys.wait(ctx) == "myapp:notifications:wait"
      assert Keys.active(ctx) == "myapp:notifications:active"
      assert Keys.job(ctx, "456") == "myapp:notifications:456"
    end
  end
end

File: golang/keys_test.go
package bullmq

import (
	"encoding/hex"
	"testing"
)

func TestKeysUseTheSharedNamingConvention(t *testing.T) {
	k, err := NewKeys("emails", "")
	if err != nil {
		t.Fatalf("NewKeys: %v", err)
	}

	cases := map[string]string{
		"base":        k.Base(),
		"keyPrefix":   k.KeyPrefix(),
		"wait":        k.Wait(),
		"active":      k.Active(),
		"delayed":     k.Delayed(),
		"prioritized": k.Prioritized(),
		"completed":   k.Completed(),
		"failed":      k.Failed(),
		"paused":      k.Paused(),
		"stalled":     k.Stalled(),
		"events":      k.Events(),
		"meta":        k.Meta(),
		"marker":      k.Marker(),
		"pc":          k.PC(),
		"id":          k.ID(),
		"job":         k.Job("42"),
		"jobLock":     k.JobLock("42"),
		"jobLogs":     k.JobLogs("42"),
		"deps":        k.Dependencies("42"),
		"metrics":     k.Metrics("completed"),
	}
	want := map[string]string{
		"base":        "bull:emails",
		"keyPrefix":   "bull:emails:",
		"wait":        "bull:emails:wait",
		"active":      "bull:emails:active",
		"delayed":     "bull:emails:delayed",
		"prioritized": "bull:emails:prioritized",
		"completed":   "bull:emails:completed",
		"failed":      "bull:emails:failed",
		"paused":      "bull:emails:paused",
		"stalled":     "bull:emails:stalled",
		"events":      "bull:emails:events",
		"meta":        "bull:emails:meta",
		"marker":      "bull:emails:marker",
		"pc":          "bull:emails:pc",
		"id":          "bull:emails:id",
		"job":         "bull:emails:42",
		"jobLock":     "bull:emails:42:lock",
		"jobLogs":     "bull:emails:42:logs",
		"deps":        "bull:emails:42:dependencies",
		"metrics":     "bull:emails:metrics:completed",
	}
	for name, got := range cases {
		if got != want[name] {
			t.Errorf("%s = %q, want %q", name, got, want[name])
		}
	}
}

func TestKeysHonourACustomPrefix(t *testing.T) {
	k, err := NewKeys("emails", "{tenant}")
	if err != nil {
		t.Fatalf("NewKeys: %v", err)
	}
	if got, want := k.Wait(), "{tenant}:emails:wait"; got != want {
		t.Errorf("Wait() = %q, want %q", got, want)
	}
}

func TestNewKeysRejectsInvalidNames(t *testing.T) {
	for _, name := range []string{"", "with:colon"} {
		if _, err := NewKeys(name, ""); err == nil {
			t.Errorf("NewKeys(%q) should have failed", name)
		}
	}
}

func TestClientNameMatchesTheNodeFormat(t *testing.T) {
	k, _ := NewKeys("emails", "")
	// base64("emails") == "ZW1haWxz"
	if got, want := k.ClientName(":w:1"), "bull:ZW1haWxz:w:1"; got != want {
		t.Errorf("ClientName() = %q, want %q", got, want)
	}
}

func TestResolveParentQueueKey(t *testing.T) {
	got, err := resolveParentQueueKey("bull", "parents")
	if err != nil || got != "bull:parents" {
		t.Fatalf("resolveParentQueueKey(bare) = %q, %v", got, err)
	}
	got, err = resolveParentQueueKey("bull", "bull:parents")
	if err != nil || got != "bull:parents" {
		t.Fatalf("resolveParentQueueKey(qualified) = %q, %v", got, err)
	}
	if _, err := resolveParentQueueKey("bull", "other:parents"); err == nil {
		t.Fatal("resolveParentQueueKey should reject a foreign prefix")
	}
}

func TestRandomIDIsUniqueAndHex(t *testing.T) {
	seen := make(map[string]struct{}, 100)
	for range 100 {
		id := randomID()
		if _, err := hex.DecodeString(id); err != nil {
			t.Fatalf("randomID() = %q, not hex: %v", id, err)
		}
		if _, dup := seen[id]; dup {
			t.Fatalf("randomID() produced a duplicate: %q", id)
		}
		seen[id] = struct{}{}
	}
}
